src/Controls/src/BindingSourceGen/BindingSourceGenerator.cs

```diff
@@ -59,54 +72,84 @@ static bool IsNullable(ITypeSymbol type)
 		return false;
 	}
 
-	static Binding? GetBindingForGeneration(GeneratorSyntaxContext context, CancellationToken t)
+	static BindingDiagnosticsWrapper GetBindingForGeneration(GeneratorSyntaxContext context, CancellationToken t)
 	{
+		var diagnostics = new List<Diagnostic>();
 		var invocation = (InvocationExpressionSyntax)context.Node;
 
 		var method = invocation.Expression as MemberAccessExpressionSyntax;
 
+		var methodSymbolInfo = context.SemanticModel.GetSymbolInfo(method!, cancellationToken: t);
 
-		var sourceCodeLocation = new SourceCodeLocation(
-			context.Node.SyntaxTree.FilePath,
-			method!.Name.GetLocation().GetLineSpan().StartLinePosition.Line + 1,
-			method!.Name.GetLocation().GetLineSpan().StartLinePosition.Character + 1
-		);
+		if (methodSymbolInfo.Symbol is not IMethodSymbol methodSymbol) //TODO: Do we need this check?
+		{
+			diagnostics.Add(Diagnostic.Create(
+				DiagnosticsDescriptors.UnableToResolvePath, method!.GetLocation()));
+			return new BindingDiagnosticsWrapper(null, diagnostics.ToArray());
+		}
 
+		// Check whether we are using correct overload
+		if (methodSymbol.Parameters.Length < 2 || methodSymbol.Parameters[1].Type.Name != "Func")
+		{
+			diagnostics.Add(Diagnostic.Create(
+				DiagnosticsDescriptors.SuboptimalSetBindingOverload, method!.GetLocation()));
+			return new BindingDiagnosticsWrapper(null, diagnostics.ToArray());
+		}
+	
 		var argumentList = invocation.ArgumentList.Arguments;
 		var getter = argumentList[1].Expression;
 
-
+		//Check if getter is a lambda
 		if (getter is not LambdaExpressionSyntax lambda)
 		{
-			return null; // TODO: Optimize
+			diagnostics.Add(Diagnostic.Create(
+				DiagnosticsDescriptors.GetterIsNotLambda, getter.GetLocation()));
+			return new BindingDiagnosticsWrapper(null, diagnostics.ToArray());
 		}
 
-
-		if (context.SemanticModel.GetSymbolInfo(lambda).Symbol is not IMethodSymbol symbol)
+		//Check if lambda body is an expression
+		if (lambda.Body is not ExpressionSyntax)
 		{
-			return null;
-		}; // TODO
+			diagnostics.Add(Diagnostic.Create(
+				DiagnosticsDescriptors.GetterLambdaBodyIsNotExpression, lambda.Body.GetLocation()));
+			return new BindingDiagnosticsWrapper(null, diagnostics.ToArray());
+		}
+
+		var lambdaSymbol = context.SemanticModel.GetSymbolInfo(lambda, cancellationToken: t).Symbol as IMethodSymbol;
 
-		var inputType = symbol.Parameters[0].Type;
+		var inputType = lambdaSymbol!.Parameters[0].Type;
```

Comment:
There a few places where we deal with things which can possibly be null but in all reasonable use case scenarios should not be null. Do we want to emit diagnostics when they happen to be null? Or just skip them

Reply:
I think we should do strict C-style casts in these cases instead of `as`.

Reply:
I changed this to C-style when casting to MemberAccessExpressionSyntax.We cannot cast 
`context.SemanticModel.GetSymbolInfo(lambda, cancellationToken: t).Symbol as IMethodSymbol` like that as `Converting null literal or possible null value to non-nullable type.`

The `Symbol` can be null if semantic model fails to resolve the method. I encountered this problem when my test project did not reference` System.Runtime.dll`

Reply:
So if you can't do the strict cast, then check for null and throw if it is null. I'd like to avoid all `!` if possible (it should be possible).

Reply:
I understand we want to avoid `!`. The question really is should we throw an error if we find that this is null or emit a diagnostic and continue to the next Binding.

Reply:
I would throw for now and consider safer options later.